packages/manager/src/features/IAM/hooks/adapters/permissionAdapters.ts

```diff
@@ -2,71 +2,138 @@ import { accountGrantsToPermissions } from './accountGrantsToPermissions';
 import { firewallGrantsToPermissions } from './firewallGrantsToPermissions';
 import { linodeGrantsToPermissions } from './linodeGrantsToPermissions';
 
-import type { AccessType, Grants, PermissionType } from '@linode/api-v4';
+import type { EntityBase } from '../usePermissions';
+import type {
+  AccessType,
+  Grants,
+  GrantType,
+  PermissionType,
+  Profile,
+} from '@linode/api-v4';
 
-export const toPermissionMap = (
-  permissionsToCheck: PermissionType[],
-  usersPermissions: PermissionType[],
-  isRestricted?: boolean
-): Record<PermissionType, boolean> => {
-  const unrestricted = isRestricted === false; // explicit === false since the profile can be undefined
-  const usersPermissionMap = {} as Record<PermissionType, boolean>;
-  usersPermissions?.forEach(
-    (permission) => (usersPermissionMap[permission] = true)
-  );
+export type EntityPermissionMap = Record<number, PermissionMap>;
 
-  const permissionMap = {} as Record<PermissionType, boolean>;
-  permissionsToCheck?.forEach(
-    (permission) =>
-      (permissionMap[permission] =
-        (unrestricted || usersPermissionMap[permission]) ?? false)
-  );
+export type PermissionMap = Record<PermissionType, boolean>;
 
-  return permissionMap;
+/** Convert entities and the permissions associated with the entity from grants */
+export const entityPermissionMapFrom = (
+  grants: Grants | undefined,
+  grantType: GrantType,
```

Comment:
Can we name this entity type instead?
```suggestion
  entityType: EntityType,
```

Reply:
Ya even though they contain the same strings, using EntityType you'll have to cast to GrantType.